cmd/openshift-install/create.go

```diff
@@ -166,7 +166,8 @@ func destroyBootstrap(ctx context.Context, directory string) (err error) {
 	}
 	defer cleanup()
 
-	logrus.Info("Waiting for bootstrap completion...")
+	bootstrapTimeout := 30 * time.Minute
+	logrus.Infof("Waiting %v for bootstrap completion...", bootstrapTimeout)
```

Comment:
This isn't quite right.  Currently we have two phases:

1. [Waiting until the API is alive enough to tell us the server version][1], which gets a 30-minute timeout.
2. [Waiting for the `bootstrap-complete` event][2], which gets a separate, 30-minute timeout.

So in some crazy world where the API comes up very fast, you could spend two minutes waiting, get the API server version, and then time out waiting for `bootstrap-complete` 32 minutes after entering `destroyBootstrap`.  Or you could spend 29 minutes waiting, get the API server version, and time out waiting for `bootstrap-complete` 59 minutes after entering `destroyBootstrap`.  In practice getting the server version and getting `bootstrap-complete` are separate enough that I'd rather have something like:

```diff
diff --git a/cmd/openshift-install/create.go b/cmd/openshift-install/create.go
index 8dee063..623e521 100644
--- a/cmd/openshift-install/create.go
+++ b/cmd/openshift-install/create.go
@@ -166,8 +166,6 @@ func destroyBootstrap(ctx context.Context, directory string) (err error) {
 	}
 	defer cleanup()
 
-	bootstrapTimeout := 30 * time.Minute
-	logrus.Infof("Waiting %v for bootstrap completion...", bootstrapTimeout)
 	config, err := clientcmd.BuildConfigFromFlags("", filepath.Join(directory, "auth", "kubeconfig"))
 	if err != nil {
 		return errors.Wrap(err, "loading kubeconfig")
@@ -180,32 +178,41 @@ func destroyBootstrap(ctx context.Context, directory string) (err error) {
 
 	discovery := client.Discovery()
 
-	apiContext, cancel := context.WithTimeout(ctx, bootstrapTimeout)
+	apiTimeout := 30 * time.Minute
+	logrus.Infof("Waiting %v for the Kubernetes API...", apiTimeout)
+	apiContext, cancel := context.WithTimeout(ctx, apiTimeout)
 	defer cancel()
-	previousError := ""
 	// Only show error messages after 30s have elapsed; before that we know
 	// that even in the fastest setups that machines are going to be
 	// booting, running Ignition, etc.
-	beginVerboseTime := time.Now().Add(30 * time.Second)
+	logDownsample := 15
+	silenceRemaining := logDownsample
+	previousErrorSuffix := ""
 	wait.Until(func() {
 		version, err := discovery.ServerVersion()
 		if err == nil {
 			logrus.Infof("API %s up", version)
 			cancel()
 		} else {
-			errText := err.Error()
-			if previousError != errText && time.Now().After(beginVerboseTime) {
-				logrus.Infof("API not up yet: %s", errText)
-				previousError = errText
-			} else {
-				logrus.Debugf("API not up yet: %s", errText)
+			silenceRemaining--
+			chunks := strings.Split(err.Error(), ":")
+			errorSuffix := chunks[len(chunks)-1]
+			if previousErrorSuffix != errorSuffix {
+				logrus.Debugf("Still waiting for the Kubernetes API: %v", err)
+				previousErrorSuffix = errorSuffix
+				silenceRemaining = logDownsample
+			} else if silenceRemaining == 0 {
+				logrus.Debugf("Still waiting for the Kubernetes API: %v", err)
+				silenceRemaining = logDownsample
 			}
 		}
-	}, 5*time.Second, apiContext.Done())
+	}, 2*time.Second, apiContext.Done())
 
 	events := client.CoreV1().Events("kube-system")
 
-	eventContext, cancel := context.WithTimeout(ctx, 30*time.Minute)
+	eventTimeout := 30 * time.Minute
+	logrus.Infof("Waiting %v for the bootstrap-complete event...", eventTimeout)
+	eventContext, cancel := context.WithTimeout(ctx, eventTimeout)
 	defer cancel()
 	_, err = Until(
 		eventContext,
```

Example output (from `${INSTALL_DIR}/.openshift_install.log` to get timestamps):

```
time="2018-12-03T21:39:10-08:00" level=info msg="Waiting 30m0s for the Kubernetes API..."
time="2018-12-03T21:39:20-08:00" level=debug msg="Still waiting for the Kubernetes API: Get https://wking-api.installer.testing:6443/version?timeout=32s: dial tcp 192.168.126.10:6443: connect: no route to host"
time="2018-12-03T21:39:22-08:00" level=debug msg="Still waiting for the Kubernetes API: Get https://wking-api.installer.testing:6443/version?timeout=32s: dial tcp 192.168.126.10:6443: connect: connection refused"
time="2018-12-03T21:39:52-08:00" level=debug msg="Still waiting for the Kubernetes API: Get https://wking-api.installer.testing:6443/version?timeout=32s: dial tcp 192.168.126.10:6443: connect: connection refused"
time="2018-12-03T21:40:22-08:00" level=debug msg="Still waiting for the Kubernetes API: Get https://wking-api.installer.testing:6443/version?timeout=32s: dial tcp 192.168.126.11:6443: connect: connection refused"
time="2018-12-03T21:40:52-08:00" level=debug msg="Still waiting for the Kubernetes API: Get https://wking-api.installer.testing:6443/version?timeout=32s: dial tcp 192.168.126.10:6443: connect: connection refused"
time="2018-12-03T21:40:54-08:00" level=debug msg="Still waiting for the Kubernetes API: the server could not find the requested resource"
time="2018-12-03T21:41:24-08:00" level=debug msg="Still waiting for the Kubernetes API: the server could not find the requested resource"
time="2018-12-03T21:41:54-08:00" level=debug msg="Still waiting for the Kubernetes API: the server could not find the requested resource"
time="2018-12-03T21:42:24-08:00" level=debug msg="Still waiting for the Kubernetes API: the server could not find the requested resource"
time="2018-12-03T21:42:54-08:00" level=debug msg="Still waiting for the Kubernetes API: the server could not find the requested resource"
time="2018-12-03T21:43:04-08:00" level=debug msg="Still waiting for the Kubernetes API: Get https://wking-api.installer.testing:6443/version?timeout=32s: dial tcp 192.168.126.11:6443: connect: connection refused"
time="2018-12-03T21:43:35-08:00" level=debug msg="Still waiting for the Kubernetes API: Get https://wking-api.installer.testing:6443/version?timeout=32s: dial tcp 192.168.126.10:6443: connect: connection refused"
time="2018-12-03T21:44:07-08:00" level=info msg="API v1.11.0+915f688 up"
time="2018-12-03T21:44:07-08:00" level=info msg="Waiting 30m0s for the bootstrap-complete event..."
```

The `server could not find the requested resource` phase is when the etcd-signer is running its dummy API.  Then the API goes away again until we get to the cluster-bootstrap (previously bootkube) API.

[1]: https://github.com/openshift/installer/blob/v0.5.0/cmd/openshift-install/create.go#L182-L192
[2]: https://github.com/openshift/installer/blob/v0.5.0/cmd/openshift-install/create.go#L194-L239

Reply:
This patch-on-my-patch LGTM overall.

Maybe what we really should do at least for interactive TTYs is be a simple curses app that allows one to easily toggle into "view details" mode?  But eh.

I actually wasn't aware of the `.openshift_install.log` file.  What happened is that I went and ssh'd to the master and was reading logs etc. - trying to debug completely the wrong thing.

Reply:
Updated with your changes :arrow_down: